### ISSUE TITLE
Parcel 2: Abstract filesystem to support more environments

### ISSUE THREAD

**Issue by devongovett** [parcel-bundler/parcel#2143]
Create a `@parcel/fs` package in `packages/core/fs` which contains promisified versions of all necessary fs functions. This is currently in the utils of Parcel 1. Update all references in Parcel 1 and Parcel 2 to use this package for all filesystem access. Plugins will also be able to use it if they need direct access to the filesystem.

Parcel 2 will support running in environments other than node, e.g. browsers. This could be accomplished by building Parcel using an alias for `@parcel/fs` to some other filesystem package.

**Comment by humphd** [comment]
This is interesting.  We've got a drop-in replacement for `fs` for browsers in https://github.com/filerjs/filer.  It would be neat to be able to run parcel in the browser.

Doing a mix of browser based fs like this, and combining it with Service Workers to cache URLs, works really well.

cc @modeswitch

**Comment by mischnic** [comment]
An in-memory fs would also be useful for the parcel test suite.

**Comment by mischnic** [comment]
> Create a @parcel/fs package in packages/core/fs which contains promisified versions of all necessary fs functions.

Isn't this already resolved?

**Comment by devongovett** [comment]
Yeah that part is done. I wanted to possibly experiment with passing an fs instance around rather than importing directly from `@parcel/fs` through. For the tests, ideally we'd want to read from the filesystem directly, but write the output into a virtual in-memory fs. For that, we'd need `inputFS` and `outputFS` options to parcel that would get passed around everywhere.

**Comment by mischnic** [comment]
Maybe an alternative to passing fs objects around:

Configuring the `@parcel/fs` package once and then using prefixes to specify input/output
```js
import fs from "@parcel/fs";
fs.setInput(...);
fs.setOutput(...);

// ...

fs.writeFile("out://dist/index.html");
fs.readdir("in://.");

**Comment by devongovett** [comment]
This is mostly done in #3244. Remaining work is to abstract the watcher so watch events work when using the MemoryFS.
